### PR TITLE
Abduction

### PR DESCRIPTION
**Pull request by Fox-McCloud**
Ports TG's abductors, see the wiki for more details: https://tgstation13.org/wiki/Abductor

Summary of game mode:
- There's a team of abductors that is created, at round-start---they're tasked with capturing individuals on station and performing experimental surgery on them by inserting biomechanical glands into them, then sending them back to the station.
- Glands to various things, from healing all the way to out and out gibbing; the effects are random and largely designed to cause chaos
- Abductors can understand all languages, but they can only speak over their abductor hivemind---only abductors on the same team as them can hear each other.
- Abductors have a number unique tools to help accomplish their goals; a baton that can stun for a long time, cuff people, and yes...even probe them (it can even detect changelings). They also have a vest that can reduce stuns or disguise themselves to look like crewmembers.
- The key piece of equipment in abductors is their camera console, which allows them to observe the crew, teleport to a set location, retrieve marked crewmembers, or switch modes on your vest.

A lot of team-work and co-ordination is needed to play this game mode--even more so than Nuke Ops; for that reason, it has the highest antag account age requirement (tied with malf).

Pic
img

Abductors aren't just a game-mode; they're a late game event that can happen around the same time that Xenomorphs could potentially happen---teams are formed from the ghost population.

Big thanks to @Nienhaus for cutting up the Abductors!

Feel free to ask any/all questions!

Things To Do (Before Review/Merging)
- [x] Test, test, and re-test!

Differences from TG:
- No image of your abductee on the dissection machine UI (zero mechanics/balance impacts)

Other 
- Refactored closet deconstruction
- Removed cloaking device 
 - fixes https://github.com/ParadiseSS13/Paradise/issues/3316 ....kinda..not really..

:cl: Fox McCloud
add: Adds in Abductors Game Mode
/:cl:

### CODE REVIEW

**Comment by Mijhead** [comment]
:+1:

**Comment by Spacemanspark** [comment]
:+1:
One question, however;
Why are they not greys?

**Comment by tigercat2000** [comment]
@Spacemanspark metafriends + better sprites

**Comment by KasparoVy** [comment]
Congrats on finally getting this going Fox, really looking forward to it :+1: :x: :100:

**Comment by Fox-McCloud** [comment]
>Congrats on finally getting this going Fox

Thanks--been wanting this to be a thing for about a year now---big thanks to @Aurorablade and @tigercat2000 for adding in various systems that laid the groundwork for this to be even remotely possible.

File: code/game/gamemodes/miniantags/abduction/abduction.dm [hunk 1]
```diff
@@ -0,0 +1,445 @@
...
+	var/obj/effect/landmark/L
+	var/datum/mind/scientist
+	var/team_name
+	var/mob/living/carbon/human/H
+
+	team_name = team_names[team_number]
+	agent = agents[team_number]
+	H = agent.current
+	L = agent_landmarks[team_number]
+	H.loc = L.loc
```
Comment (tigercat2000) [inline code/game/gamemodes/miniantags/abduction/abduction.dm]: H.forceMove(L.loc)

**Comment by Spacemanspark** [comment]
Oh, wait, one other thing.
Are we also getting the EMP gland? The one that randomly shoots out EMPs? I don't care either way, but I just wanted to know beforehand.

File: code/game/gamemodes/miniantags/abduction/abduction_gear.dm [hunk 2]
```diff
@@ -0,0 +1,634 @@
...
+			return
+
+/obj/item/clothing/suit/armor/abductor/vest/proc/SetDisguise(datum/icon_snapshot/entry)
+	disguise = entry
+
+/obj/item/clothing/suit/armor/abductor/vest/proc/ActivateStealth()
+	if(disguise == null)
+		return
+	stealth_active = 1
+	if(istype(src.loc, /mob/living/carbon/human))
```
Comment (tigercat2000) [inline code/game/gamemodes/miniantags/abduction/abduction_gear.dm]: ishuman(loc)

File: code/game/gamemodes/miniantags/abduction/abduction_gear.dm [hunk 3]
```diff
@@ -0,0 +1,634 @@
...
+		M.overlays = disguise.overlays
+		M.update_inv_r_hand()
+		M.update_inv_l_hand()
+	return
+
+/obj/item/clothing/suit/armor/abductor/vest/proc/DeactivateStealth()
+	if(!stealth_active)
+		return
+	stealth_active = 0
+	if(istype(src.loc, /mob/living/carbon/human))
```
Comment (tigercat2000) [inline code/game/gamemodes/miniantags/abduction/abduction_gear.dm]: ishuman(loc)

File: code/game/gamemodes/miniantags/abduction/abduction_gear.dm [hunk 4]
```diff
@@ -0,0 +1,634 @@
...
+		if(VEST_COMBAT)
+			Adrenaline()
+		if(VEST_STEALTH)
+			if(stealth_active)
+				DeactivateStealth()
+			else
+				ActivateStealth()
+
+/obj/item/clothing/suit/armor/abductor/vest/proc/Adrenaline()
+	if(istype(src.loc, /mob/living/carbon/human))
```
Comment (tigercat2000) [inline code/game/gamemodes/miniantags/abduction/abduction_gear.dm]: ishuman(loc)

**Comment by Fox-McCloud** [comment]
@Spacemanspark 

All glands are implemented---every single one, from the healing one, all the way up to the one that makes an SNPC.

This is pretty much as close to a 1:1 port as you can get.

File: code/game/gamemodes/miniantags/abduction/abduction_gear.dm [hunk 5]
```diff
@@ -0,0 +1,634 @@
...
+
+/obj/item/device/abductor/silencer/afterattack(atom/target, mob/living/user, flag, params)
+	if(flag)
+		return
+	if(!AbductorCheck(user))
+		return
+	radio_off(target, user)
+
+/obj/item/device/abductor/silencer/proc/radio_off(atom/target, mob/living/user)
+	if( !(user in (viewers(7,target))) )
```
Comment (tigercat2000) [inline code/game/gamemodes/miniantags/abduction/abduction_gear.dm]: the spacing in this actually hurts

**Comment by Spacemanspark** [comment]
Nice.

File: code/game/gamemodes/miniantags/abduction/abduction.dm [hunk 6]
```diff
@@ -0,0 +1,445 @@
...
+	var/obj/effect/landmark/L
+	var/datum/mind/scientist
+	var/team_name
+	var/mob/living/carbon/human/H
+	for(var/team_number=1,team_number<=abductor_teams,team_number++)
+		team_name = team_names[team_number]
+		agent = agents[team_number]
+		H = agent.current
+		L = agent_landmarks[team_number]
+		H.loc = L.loc
```
Comment (tigercat2000) [inline code/game/gamemodes/miniantags/abduction/abduction.dm]: forcemove on pretty much any of these

**Comment by Fox-McCloud** [comment]
Also to those who are going to ask "why not use the Grey's sprite for this?"

Simple, I wanted to avoid "Vox Raider" syndrome. where the on-station crewmembers who are that species instantly ally themselves (even if Captain/HoS or previous loyalty) with the enemy just because "they are us!". No abductors aren't your friends; they're an EoC and they don't care about any species other than themselves--they even dislike each other.

**Comment by NTSAM** [comment]
ayy lmao ✌️ 👽 

👍 finally, a fresh new roundtype.

**Comment by QuinnAggeler** [comment]
All the yes. Ever since I heard of this I wanted to try it. Can't wait to see it in action!

**Comment by FlattestGuitar** [comment]
I'm pretty sure they look too similar to Grey. How about we resprite them to be more green? This will help avoid the "Voxxy is kin" syndrome, as well as help out new players in understanding what they're not getting killed by.

**Comment by Fox-McCloud** [comment]
Corrected some `loc` setting in favor of `forceMove`, ensured mob icons update after abductors are equipped, and allowed people without organs (minus IPCs) to be able to be dissected too.

**Comment by Fox-McCloud** [comment]
Vest disguises now work.

It's the most bullshit snowflake code, ever, but our human `update_icons` code is the most frustrating and inane thing to work with, and there's really no other way that's not equally as bad.

Removed the cloaking device (which was previously snowflaked into human `update_icons`) because there can only be one-----snowflake.

**Comment by Keilden** [comment]
So the wiki does not say what the body snatchers are, are they SNPCs?

**Comment by Fox-McCloud** [comment]
>So the wiki does not say what the body snatchers are, are they SNPCs?

Yup!

**Comment by monster860** [comment]
Souns fun.

File: code/game/area/Space Station 13 areas.dm [hunk 7]
```diff
@@ -618,6 +618,14 @@ var/list/ghostteleportlocs = list()
 	name = "\improper Hyperspace"
 	icon_state = "shuttle"
 
+//Abductors
+/area/abductor_ship
+	name = "\improper Abductor Ship"
+	icon_state = "yellow"
```
Comment (monster860) [inline code/game/area/Space Station 13 areas.dm]: Yellow: The only antag area icon state

**Comment by Fox-McCloud** [comment]
Implemented the abductor tech to R&D, in addition to the abductor walls.

**Comment by Fox-McCloud** [comment]
Fixed up several things and some quirks with the camera console (should make Xenobio console behave better, too).

Tested with @DaveTheHeadcrab @Aurorablade and Jay---everything works properly--getting two runtimes, though I believe they're related to not having a database set up---once I can confirm that, then this is ready to merge.

File: code/game/area/Space Station 13 areas.dm [hunk 8]
```diff
@@ -618,6 +618,14 @@ var/list/ghostteleportlocs = list()
 	name = "\improper Hyperspace"
 	icon_state = "shuttle"
 
+//Abductors
+/area/abductor_ship
+	name = "\improper Abductor Ship"
+	icon_state = "yellow"
+	requires_power = 0
+	has_gravity = 1
+	lighting_use_dynamic = 0
```
Comment (tigercat2000) [inline code/game/area/Space Station 13 areas.dm]: grumble gramble, should use fancy colored lights for mood

File: code/game/gamemodes/miniantags/abduction/abduction.dm [hunk 9]
```diff
@@ -0,0 +1,456 @@
...
+		var/datum/objective/objective = team_objectives[team_number]
+		var/team_name = team_names[team_number]
+		if(console.experiment.points >= objective.target_amount)
+			to_chat(world, "<span class='greenannounce'>[team_name] team fullfilled its mission!</span>")
+		else
+			to_chat(world, "<span class='boldannounce'>[team_name] team failed its mission.</span>")
+	..()
+	return 1
+
+/datum/game_mode/proc/auto_declare_completion_abduction()
```
Comment (tigercat2000) [inline code/game/gamemodes/miniantags/abduction/abduction.dm]: can i just note how bad these auto_declare_completion procs are
like why does gamemode code do this
why not use an object oriented standard like auto_declare_completion() defined on each path
aaaaaaaaaahhhhhhhhhhhh

File: code/game/gamemodes/miniantags/abduction/abduction.dm [hunk 10]
```diff
@@ -0,0 +1,456 @@
+/datum/game_mode
```
Comment (tigercat2000) [inline code/game/gamemodes/miniantags/abduction/abduction.dm]: the `abduction/` folder should really be in `gamemodes/` not `miniantags/`

Comment (Fox-McCloud) [inline code/game/gamemodes/miniantags/abduction/abduction.dm]: while it's a game-mode, in theory, in practice it's far more like an event----generally speaking, that's what players will know of it, too.

The event utilizes a lot of the game-modes procs for equipping/setting up things.

File: code/game/gamemodes/miniantags/abduction/abduction.dm (hunk 9 above)
Comment (Fox-McCloud) [inline code/game/gamemodes/miniantags/abduction/abduction.dm]: because game mode code is ancient and bad and makes everyone feel bad.

**Comment by Fox-McCloud** [comment]
All bugs/runtimes (that I could fine) should be fixed.

Big thanks to @Tastyfish for helping out with the last two remaining runtimes and fixing them.

**Comment by ProperPants** [comment]
Is the abductor stuff gonna have those sprites? They look different from ours, but I suppose if they remain mostly contained to the abductor shuttle, it's not a big deal.

**Comment by Dumbdumn5** [comment]
:+1: Because DAMN, THOSE GLANDS LOOK AWESOME!

File: code/game/gamemodes/miniantags/abduction/abduction.dm [hunk 11]
```diff
@@ -0,0 +1,456 @@
...
+		obj_count++
+	return
+
+/datum/game_mode/abduction/proc/greet_scientist(datum/mind/abductor,team_number)
+	abductor.objectives += team_objectives[team_number]
+	var/team_name = team_names[team_number]
+
+	to_chat(abductor.current, "<span class='notice'>You are a scientist of [team_name]!</span>")
+	to_chat(abductor.current, "<span class='notice'>With the help of your teammate, kidnap and experiment on station crew members!</span>")
+	to_chat(abductor.current, "<span class='notice'>Use your tool and ship consoles to support the agent and retrieve human specimens.</span>")
```
Comment (KasparoVy) [inline code/game/gamemodes/miniantags/abduction/abduction.dm]: Only humans?

File: code/game/gamemodes/miniantags/abduction/abduction.dm (hunk 9 above)
Comment (KasparoVy) [inline code/game/gamemodes/miniantags/abduction/abduction.dm]: makes us cringe every time
or cry
occasionally both

File: code/game/gamemodes/miniantags/abduction/machinery/camera.dm [hunk 12]
```diff
@@ -0,0 +1,165 @@
+/obj/machinery/computer/camera_advanced/abductor
+	name = "Human Observation Console"
```
Comment (KasparoVy) [inline code/game/gamemodes/miniantags/abduction/machinery/camera.dm]: Only humans?

File: code/game/gamemodes/miniantags/abduction/machinery/pad.dm [hunk 13]
```diff
@@ -0,0 +1,54 @@
+/obj/machinery/abductor/pad
+	name = "Alien Telepad"
+	desc = "Use this to transport to and from human habitat"
```
Comment (KasparoVy) [inline code/game/gamemodes/miniantags/abduction/machinery/pad.dm]: Only humans?

Comment (tigercat2000) [inline code/game/gamemodes/miniantags/abduction/machinery/pad.dm]: @KasparoVy they're unintegrated ayyliens, they probably view everything as different breeds of humans
that's why they probe them

**Pull request merged**